### PR TITLE
Clusterinfo consistency

### PR DESCRIPTION
**Pull request by stevenzzzz**
*Description*: 
Cache threadlocalCluster.info in ActiveStream and open it to filters via StreamFilterCallbacks.

*Risk Level*: Low (nobody uses it).
*Testing*: unit tests.
*Docs Changes*:
*Release Notes*:
[Optional Fixes #Issue] 4558
[Optional *Deprecated*:]

### CODE REVIEW

**Comment by stevenzzzz** [comment]
@mattklein123 Hey Matt PTAL and see if it's okay. (I made it to be always on as I don't think it should be off when we have the API).

File: include/envoy/http/filter.h [hunk 1]
```diff
@@ -111,6 +112,13 @@ class StreamFilterCallbacks {
    */
   virtual Router::RouteConstSharedPtr route() PURE;
 
+  /**
+   * Returns the clusterInfo for the current request.
+   * This method is to avoid multiple look ups in the filter chain, it also provides a consistent
+   * view of clusterInfo after a route is picked/repicked.
+   * NOTE: Cached clusterInfo and route should be updated the same time.
```
[the remark this answers is not in the document]
Comment (stevenzzzz) [reply include/envoy/http/filter.h]: done

File: include/envoy/http/filter.h [hunk 2]
```diff
@@ -111,6 +112,13 @@ class StreamFilterCallbacks {
    */
   virtual Router::RouteConstSharedPtr route() PURE;
 
+  /**
+   * Returns the clusterInfo for the current request.
```
[the remark this answers is not in the document]
Comment (stevenzzzz) [reply include/envoy/http/filter.h]: done

**Comment by stevenzzzz** [comment]
Hey Matt, I want to make the switch in another PR if that's fine by you?

**Comment by mattklein123** [comment]
> Hey @mattklein123 Matt, I want to make the switch in another PR if that's fine by you?

Do you promise you will do a follow up PR right after this one?

Can you please also merge master?

**Comment by stevenzzzz** [comment]
sure thing. :)
I found there is another similar cache route implementation, but at network filter layer: https://github.com/envoyproxy/envoy/blob/master/source/extensions/filters/network/thrift_proxy/conn_manager.cc
I think we want to do the same thing there?

**Comment by mattklein123** [comment]
> I think we want to do the same thing there?

cc @zuercher ^. Yeah we likely do but I wouldn't worry about it right now.

**Comment by stevenzzzz** [comment]
wow, so all I did is:
git checkout master 
git pull upstream master
git checkout feature-branch
git merge master <resolve conflicts>
git commit
git push origin

Annnd there are 160+ files in my PR. Which I think most changes are in sync with master HEAD but shown as diff here.
Sorry for the mess, I need some git help to figure out what happened.

**Comment by mattklein123** [comment]
@stevenzzzz please check CI

File: source/common/http/conn_manager_impl.cc [hunk 3]
```diff
@@ -1477,8 +1484,17 @@ Tracing::Span& ConnectionManagerImpl::ActiveStreamFilterBase::activeSpan() {
 
 Tracing::Config& ConnectionManagerImpl::ActiveStreamFilterBase::tracingConfig() { return parent_; }
 
+Upstream::ClusterInfoConstSharedPtr ConnectionManagerImpl::ActiveStreamFilterBase::clusterInfo() {
+  // NOTE: Refreshing route caches clusterInfo as well.
+  if (!parent_.cached_route_.has_value()) {
```
[the remark this answers is not in the document]
Comment (stevenzzzz) [reply source/common/http/conn_manager_impl.cc]: ack

File: include/envoy/http/filter.h [hunk 4]
```diff
@@ -111,6 +112,13 @@ class StreamFilterCallbacks {
    */
   virtual Router::RouteConstSharedPtr route() PURE;
 
+  /**
+   * Returns the clusterInfo for the cached route.
+   * This method is to avoid multiple look ups in the filter chain, it also provides a consistent
+   * view of clusterInfo after a route is picked/repicked.
+   * NOTE: Cached clusterInfo and route will be updated the same time.
+   */
+  virtual Upstream::ClusterInfoConstSharedPtr clusterInfo() PURE;
```
Comment (mattklein123) [inline include/envoy/http/filter.h]: nit: newline after this line

Comment (stevenzzzz) [reply include/envoy/http/filter.h]: done. Thought fix_format would catch this.

**Pull request merged**